app/models/host/managed_extensions.rb

```diff
@@ -44,11 +45,15 @@ def setKexec
     template = provisioning_template(:kind => 'kexec')
     raise ::Foreman::Exception.new(N_("Kexec template not associated with operating system")) unless template
     @host = self
-    @kexec_kernel, @kexec_initrd = operatingsystem.boot_files_uri(@host.medium, @host.architecture)
-    json = unattended_render(template)
-    old.becomes(Host::Discovered).kexec json
+    @kexec_kernel, @kexec_initrd = operatingsystem.boot_files_uri(@host.medium, @host.architecture, @host)
+    # try to parse JSON and error out early
+    json = JSON.parse(unattended_render(template))
+    ::Foreman::Exception.new(N_("Kernel kexec URL is invalid: '%s'"), json['kernel']) unless json['kernel'] =~ /\Ahttp.+\Z/
+    ::Foreman::Exception.new(N_("Init RAM kexec URL is invalid: '%s'"), json['initrd']) unless json['initrd'] =~ /\Ahttp.+\Z/
+    old.becomes(Host::Discovered).kexec json.to_s
     true
   rescue ::Foreman::Exception
```

Comment:
In order to log the exception this line would need to be `::Foreman::Exception => e`